### PR TITLE
[apiRefactor] InterpolatedColorScale changes

### PR DESCRIPTION
**Pull request by rcchen**
* `private static _getD3InterpolatedScale(colors: string[], scaleType: string)` -> `private _getD3InterpolatedScale()`
* `private static _interpolateColors(colors: string[])` -> `private _interpolateColors()`

### CODE REVIEW

File: src/scales/interpolatedColorScale.ts [hunk 1]
```diff
@@ -135,10 +108,10 @@ export module Scales {
      *     (linear/pow/log/sqrt). Default is "linear". @see {@link scaleType}
      *     for further options.
      */
-    constructor(colorRange: any = "reds", scaleType: string = "linear") {
```
Comment (jtlan) [inline src/scales/interpolatedColorScale.ts]: mfw `any`. Can we also take this opportunity to move the `constructor` to the beginning of the file, where devs might be expected to look for it?

File: src/scales/interpolatedColorScale.ts [hunk 2]
```diff
@@ -168,29 +141,29 @@ export module Scales {
...
-    public scaleType(scaleType?: string): any {
-      if (scaleType == null) {
-        return this._scaleType;
-      }
-      this._scaleType = scaleType;
-      this._resetScale();
-      return this;
-    }
+     public colorScale(colorScale: D3.Scale.QuantitativeScale): InterpolatedColor;
+     public colorScale(colorScale?: D3.Scale.QuantitativeScale): any {
```
Comment (jtlan) [inline src/scales/interpolatedColorScale.ts]: I don't think this is the best name for this method. `internalScale()`?

File: test/scales/scaleTests.ts [hunk 3]
```diff
@@ -406,8 +406,7 @@ describe("Scales", () => {
       assert.strictEqual("#000000", scale.scale(0));
       assert.strictEqual("#ffffff", scale.scale(16));
       assert.strictEqual("#777777", scale.scale(8));
-
-      scale.scaleType("log");
+  	  scale.colorScale(d3.scale.log());
```
Comment (jtlan) [inline test/scales/scaleTests.ts]: Why is it indented?

**Comment by jtlan** [comment]
[-1] Questions around the name of the method used to set the internal scale.

Is it even necessary to have the ability to change the internal scale?

File: src/scales/interpolatedColorScale.ts (hunk 1 above)
Comment (rcchen) [inline src/scales/interpolatedColorScale.ts]: State of the union circa 9 months ago: not implemented (wow unintentional pun)

File: test/scales/scaleTests.ts (hunk 3 above)
Comment (bluong) [inline test/scales/scaleTests.ts]: There's a tab character!

File: src/scales/interpolatedColorScale.ts [hunk 4]
```diff
@@ -168,29 +141,29 @@ export module Scales {
     }
 
     /**
-     * Gets the internal scale type.
-     *
-     * @returns {string} The current scale type.
+     * Gets the internal scale.
+     *
```
Comment (bluong) [inline src/scales/interpolatedColorScale.ts]: Remove trailing whitespace

File: src/scales/interpolatedColorScale.ts [hunk 5]
```diff
@@ -168,29 +141,29 @@ export module Scales {
...
-    }
+     public colorScale(colorScale: D3.Scale.QuantitativeScale): InterpolatedColor;
+     public colorScale(colorScale?: D3.Scale.QuantitativeScale): any {
+       if (colorScale == null) {
+         return this._colorScale;
+       }
+       this._colorScale = colorScale;
+       this._resetScale();
+       return this;
+     }
```
Comment (bluong) [inline src/scales/interpolatedColorScale.ts]: Indentation is 1 extra than what it should be

File: src/scales/interpolatedColorScale.ts [hunk 6]
```diff
@@ -56,49 +56,44 @@ export module Scales {
...
-     *
-     * @param {string[]} colors an array of strings representing color
-     *     values in hex ("#FFFFFF") or keywords ("white").
-     * @returns {D3.Transition.Interpolate} The d3 interpolator for colors.
+     * Generates the converted QuantitativeScale.
+     * 
+     * @returns {D3.Scale.QuantitativeScale} The converted d3 QuantitativeScale
      */
-    private static _interpolateColors(colors: string[]): D3.Transition.Interpolate {
+    private _getD3InterpolatedScale(): D3.Scale.QuantitativeScale {
```
Comment (jtlan) [inline src/scales/interpolatedColorScale.ts]: No `get`.

**Comment by jtlan** [comment]
[minor] remove the `get` from the internal method name.

**Comment by jtlan** [comment]
Please fix the overlaying tests too.

File: src/scales/interpolatedColorScale.ts [hunk 7]
```diff
@@ -56,49 +56,56 @@ export module Scales {
...
+     * @param {string | string[]} colors an array of strings representing color
+     *     values in hex ("#FFFFFF") or keywords ("white"). Default is "reds"
      * @param {string} scaleType a string representing the underlying scale
-     *     type ("linear"/"log"/"sqrt"/"pow")
+     *     type ("linear"/"log"/"sqrt"/"pow"). Defaults to "linear"
      * @returns {D3.Scale.QuantitativeScale} The converted QuantitativeScale d3 scale.
      */
-    private static _getD3InterpolatedScale(colors: string[], scaleType: string): D3.Scale.QuantitativeScale {
-      var scale: D3.Scale.QuantitativeScale;
+    constructor(colorRange: string | string[] = "reds", scaleType = "linear") {
```
Comment (jtlan) [inline src/scales/interpolatedColorScale.ts]: Can we make this take only a `string[]`, and then have the default arrays of colors stored as `static` fields?

**Comment by jtlan** [comment]
`constructor` argument type.

File: src/scales/interpolatedColorScale.ts [hunk 8]
```diff
@@ -56,49 +56,56 @@ export module Scales {
...
-     *     values in hex ("#FFFFFF") or keywords ("white").
+     *     values in hex ("#FFFFFF") or keywords ("white"). Default is ["reds"]
      * @param {string} scaleType a string representing the underlying scale
-     *     type ("linear"/"log"/"sqrt"/"pow")
+     *     type ("linear"/"log"/"sqrt"/"pow"). Defaults to "linear"
      * @returns {D3.Scale.QuantitativeScale} The converted QuantitativeScale d3 scale.
      */
-    private static _getD3InterpolatedScale(colors: string[], scaleType: string): D3.Scale.QuantitativeScale {
-      var scale: D3.Scale.QuantitativeScale;
+    constructor(colorRange = ["reds"], scaleType = "linear") {
```
Comment (jtlan) [inline src/scales/interpolatedColorScale.ts]: What? I meant something like:
```Typescript
var l = new InterpolatedColorScale(InterpolatedColorScale.REDS);
```

File: src/scales/interpolatedColorScale.ts [hunk 9]
```diff
@@ -12,7 +11,7 @@ export module Scales {
    * By default it generates a linear scale internally.
    */
   export class InterpolatedColor extends Scale<number, string> {
-    private static _COLOR_SCALES: ColorGroups = {
+    public static COLOR_SCALES = {
```
Comment (jtlan) [inline src/scales/interpolatedColorScale.ts]: This should probably be named `COLOR_RANGES` to match up with what it's called in the constructor. Maybe the fields inside should be all-caps too.

Alternatively, just make each range a separate `static` field.

File: test/scales/scaleTests.ts [hunk 10]
```diff
@@ -338,7 +338,7 @@ describe("Scales", () => {
     });
 
     it("linearly interpolates colors in L*a*b color space", () => {
-      var scale = new Plottable.Scales.InterpolatedColor("reds");
+      var scale = new Plottable.Scales.InterpolatedColor(Plottable.Scales.InterpolatedColor.COLOR_SCALES["reds"]);
```
Comment (jtlan) [inline test/scales/scaleTests.ts]: Doesn't it default to this?

**Comment by jtlan** [comment]
[-1] Naming of the field.

**Pull request merged**